pkg/noun/v2/manage.c

```diff
@@ -0,0 +1,177 @@
+
+  for (; box_w < end_w
+         ; box_w += box_u->siz_w
+         , box_u = (void*)box_w)
+    {
+      if (!box_u->use_w)
+        continue;
+      u3_assert(box_u->siz_w);
+      u3_assert(box_u->use_w);
+      box_w[box_u->siz_w - 1] = new_p;
+      new_p = c3_align(new_p + box_u->siz_w, 2, C3_ALGHI);
+    }
+}
+
+static void
+_migrate_rewrite()
+{
+  fprintf(stderr, "loom: migration rewrite\r\n");
+
+  _cm_pack_rewrite();
+}
+
+static void
+_migrate_move(u3a_v2_road *rod_u)
+{
+  fprintf(stderr, "loom: migration move\r\n");
+
+  c3_z hiz_z = u3a_v2_heap(rod_u) * sizeof(c3_w);
+
+  /* calculate required shift distance to prevent write head overlapping read head */
+  c3_w  off_w = 1;  /* at least 1 word because u3R_v2->rut_p migrates from 1 to 2 */
+  for (u3a_v2_box *box_u = u3a_v2_into(rod_u->rut_p)
+         ; (void *)box_u < u3a_v2_into(rod_u->hat_p)
+         ; box_u = (void *)((c3_w *)box_u + box_u->siz_w))
+    off_w += box_u->siz_w & 1; /* odd-sized boxes are padded by one word */
+
+  /* shift */
+  memmove(u3a_v2_into(u3H_v2->rod_u.rut_p + off_w),
+          u3a_v2_into(u3H_v2->rod_u.rut_p),
+          hiz_z);
+  /* manually zero the former rut */
+  *(c3_w *)u3a_v2_into(rod_u->rut_p) = 0;
+
+  /* relocate boxes to DWORD-aligned addresses stored in trailing size word */
+  c3_w *box_w = u3a_v2_into(rod_u->rut_p + off_w);
+  c3_w *end_w = u3a_v2_into(rod_u->hat_p + off_w);
+  u3a_v2_box *old_u = (void *)box_w;
+  c3_w siz_w = old_u->siz_w;
+  u3p(c3_w) new_p = rod_u->rut_p + 1 + c3_wiseof(u3a_v2_box);
+  c3_w *new_w;
+
+  for (; box_w < end_w
+         ; box_w += siz_w
+         , old_u = (void *)box_w
+         , siz_w = old_u->siz_w) {
+    old_u->use_w &= 0x7fffffff;
+
+    if (!old_u->use_w)
+      continue;
+
+    new_w = (void *)u3a_v2_botox(u3a_v2_into(new_p));
+    u3_assert(box_w[siz_w - 1] == new_p);
+    u3_assert(new_w <= box_w);
+
+    c3_w i_w;
+    for (i_w = 0; i_w < siz_w - 1; i_w++)
+      new_w[i_w] = box_w[i_w];
+
+    if (siz_w & 1) {
+      new_w[i_w++] = 0;         /* pad odd sized boxes */
+      new_w[i_w++] = siz_w + 1; /* restore trailing size word */
+      new_w[0] = siz_w + 1;     /* and the leading size word */
+    }
+    else {
+      new_w[i_w++] = siz_w;
+    }
+
+    new_p += i_w;
+  }
+
+  /* restore proper heap state */
+  rod_u->rut_p = 2;
+  rod_u->hat_p = new_p - c3_wiseof(u3a_v2_box);
+
+  /* like |pack, clear the free lists and cell allocator */
+  for (c3_w i_w = 0; i_w < u3a_v2_fbox_no; i_w++)
+    u3R_v2->all.fre_p[i_w] = 0;
+
+  u3R_v2->all.fre_w = 0;
+  u3R_v2->all.cel_p = 0;
+}
+
+
+/* u3m_v2_migrate: perform loom migration if necessary.
+*/
+void
+u3m_v2_migrate()
+{
+  c3_w len_w = u3C_v2.wor_i - 1;
+  c3_w ver_w = *(u3_Loom + len_w);
+
+  u3_assert( U3V_VER1 == ver_w );
+
+  c3_w* mem_w = u3_Loom + 1;
+  c3_w  siz_w = c3_wiseof(u3v_v2_home);
+  c3_w* mat_w = (mem_w + len_w) - siz_w;
+
+  u3H_v2 = (void *)mat_w;
+  u3R_v2 = &u3H_v2->rod_u;
```

Comment:
This should set the v1 pointers. They happen to be aliased to v2, but that's an implementation detail. Conceptually, we're initializing a v1 loom here.